plugin/src/main/java/org/opensearch/ml/action/handler/MLSearchHandler.java

```diff
@@ -66,39 +70,40 @@ public MLSearchHandler(Client client, NamedXContentRegistry xContentRegistry, Mo
     public void search(SearchRequest request, ActionListener<SearchResponse> actionListener) {
         User user = RestActionUtils.getUserContext(client);
         ActionListener<SearchResponse> listener = wrapRestActionListener(actionListener, "Fail to search model version");
-        if (modelAccessControlHelper.skipModelAccessControl(user)) {
-            try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            if (modelAccessControlHelper.skipModelAccessControl(user)) {
                 client.search(request, listener);
-            } catch (Exception e) {
-                log.error(e.getMessage(), e);
-                actionListener.onFailure(e);
-            }
-        } else {
-            SearchSourceBuilder sourceBuilder = modelAccessControlHelper.createSearchSourceBuilder(user);
-            SearchRequest modelGroupSearchRequest = new SearchRequest();
-            sourceBuilder.fetchSource(new String[] { MLModelGroup.MODEL_GROUP_ID_FIELD, }, null);
-            sourceBuilder.size(10000);
-            modelGroupSearchRequest.source(sourceBuilder);
-            modelGroupSearchRequest.indices(CommonValue.ML_MODEL_GROUP_INDEX);
-            ActionListener<SearchResponse> modelGroupSearchActionListener = ActionListener.wrap(r -> {
-                if (Optional.ofNullable(r).map(SearchResponse::getHits).map(SearchHits::getTotalHits).map(x -> x.value).orElse(0L) > 0) {
-                    List<String> modelGroupIds = new ArrayList<>();
-                    Arrays.stream(r.getHits().getHits()).forEach(hit -> { modelGroupIds.add(hit.getId()); });
-
-                    request.source().query(rewriteQueryBuilder(request.source().query(), modelGroupIds));
-                    client.search(request, listener);
-                } else {
-                    log.debug("No model group found");
-                    request.source().query(rewriteQueryBuilder(request.source().query(), null));
-                    client.search(request, listener);
-                }
-            }, e -> {
-                log.error("Fail to search model groups!", e);
-                actionListener.onFailure(e);
-            });
-            try (ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
+            } else if (!clusterService.state().metadata().hasIndex(CommonValue.ML_MODEL_GROUP_INDEX)) {
+                request.source().query(rewriteQueryBuilder(request.source().query(), null));
```

Comment:
If no model group index, why need to rewrite query?

Reply:
`rewriteQueryBuilder` method is not easy to understand. Add some comments?

Reply:
Make sense, removed the rewriting.